README.md

```diff
@@ -1,3 +1,138 @@
-# elliptic-curve-solidity
+# elliptic-curve-solidity [![](https://travis-ci.com/witnet/elliptic-curve-solidity.svg?branch=master)](https://travis-ci.com/witnet/elliptic-curve-solidity)
 
-WIP
\ No newline at end of file
+`elliptic-curve-solidity` is an open source implementation of Elliptic Curve arithmetic operations written in Solidity.
+
+_DISCLAIMER: This is experimental software. **Use it at your own risk**!_
+
+The solidity library has been generalized in order to support any elliptic curve based on prime numbers up to 256 bits. It provides functions for:
+
+- Modular
+  - inverse
+  - exponentiation
+- Jacobian coordinates
+  - addition
+  - double
+  - multiplication
+- Affine coordinates
+  - inverse
+  - addition
+  - subtraction
+  - multiplication
+- Auxiliary
+  - conversion to affine coordinates
+  - derive coordinate Y from compressed ec point
+  - check if ec point is on curve
+
+`elliptic-curve-solidity` has been designed as a library with **only pure functions** aiming at decreasing gas consumption as much as possible.
+A modified library with a constructor for setting the curve parameters can be found at the branch `ref/constructor`.
+Additionally, gas consumption comparison can be found in [gas report][benchmark].
+
+## Supported curves
+
+The `elliptic-curve-solidity` contract supports up to 256-bit curves. However, it has been extensively tested for the following curves:
+
+- `secp256k1`
+- `secp256r1`
+- `secp192r1`
+- `secp224r1`
+
+Known limitations:
+
+- `deriveY` function does work with curve `secp224r1` because of the selected derivation algorithm. This curve computations are done with a modulo prime `p` such as `p=1  mod 4`, so that more complex algorithm is required (e.g. *Tonelli-Shanks algorithm*). Note that `deriveY` is just an auxiliary function, and thus does not limit the functionality of curve arithmetic operations.
```

Comment:
```suggestion
- `deriveY` function does work with curve `secp224r1` because of the selected derivation algorithm. The computations for this curve are done with a modulo prime `p` such as `p=1  mod 4`, thus a more complex algorithm is required (e.g. *Tonelli-Shanks algorithm*). Note that `deriveY` is just an auxiliary function, and thus does not limit the functionality of curve arithmetic operations.
```